hie-base/Haskell/Ide/Engine/PluginTypes.hs

```diff
@@ -1,3 +1,7 @@
+{-# LANGUAGE InstanceSigs #-}
+{-# LANGUAGE UndecidableInstances #-}
```

Comment:
I just pushed a commit creating two separate modules, one for the singletons, which actually ended up not needing `UndecidableInstances` but instead there is a different issue since `singletons` creates some type synonyms which we don’t use so they create unused warnings. To work around this this module simply exports everything. The other module does actually enable `UndecidableInstances` since there is no way around it with nested type families (which are required here, at least if we want to have somewhat clean code). I also added notes about the singletons in the haddocks of the non singletons types.